### ISSUE TITLE
Very large documents cannot be uploaded

### ISSUE THREAD

**Issue by coderbot16** [eikek/docspell#1819]
I have a very large 130 MB PDF file that I would like to place into Docspell, but I have been unsuccessful in uploading it. In all cases, timeouts kill the upload before it has time to complete. This is even the case when using `dsc` from the same server running Docspell. I feel like the timeouts should probably be relaxed here.

> Cmd: Watch: Uploading failed: An http error occurred: An error was received from http://docspell-restserver:7880/api/v1/open/integration/item/xxx: error sending request for url (http://docspell-restserver:7880/api/v1/open/integration/item/xxx): operation timed out

If it is important, I'm still using the default Postgres storage backend. I can't find anything in the documentation that would recommend against using the database backend for these large files, but if that is the case then I'd appreciate any guidance!

**Comment by coderbot16** [comment]
Update: I just observed some confusing behavior. I went back a while later and checked on the dashboard and this large file that caused `dsc` to log a timeout was sitting right there. I'm not sure how `dsc` works but evidently `dsc` still uploaded the file despite reporting that the request failed due to a timeout, which is different from what I observed through the Web UI where a file upload failure is final. Maybe that helps?

**Comment by ahanselka** [comment]
I have experienced a similar issue with both the web ui and `dsc`, however in my case the file never appeared and seemed to just never be uploaded. I can't find anywhere to set a timeout that might prevent this from happening, but so far no luck. 

In my case I was trying to upload a `.zip` of a bunch of files. I ended up having to just upload them in smaller groups.

**Comment by eikek** [comment]
Unfortunately, big files are not really supported officially (yet). However, the server is capable of streaming large payloads (it keeps up to 10M in memory per multipart-part and then writes into a temp file). But the dsc client doesn't support this atm. You could try `curl` with `--header "Transfer-Encoding: chunked`. But … I never really tested all this. Could be that there is some part that still reads everything in memory. With a decent heap setting it would still work of course. I thought that processing really large files is not very useful, but uploading a zip with many documents is a very valid use case!

**Comment by madduck** [comment]
The `dsc-consumedir` container running `dsc watch` just imported a 350 Mb file I used for testing just fine. Note that I am using the filesystem backend for file storage because we will have lots of huge files, and while I love PostgreSQL, file storage is still better done in the file system ;)

**Comment by eikek** [comment]
> The `dsc-consumedir` container running `dsc watch` just imported a 350 Mb file I used for testing just fine. Note that I am using the filesystem backend for file storage because we will have lots of huge files, and while I love PostgreSQL, file storage is still better done in the file system ;)

That's good to hear 😄 ! I think as long as you have enough memory, things should work. It can be improved by proper streaming the data. But to be honest, for me it's not really important to support files of several GB size. This then also requires a different take on OCR etc.

On another note,I don't think you need to be afraid of PostgreSQL storing large files. They are organized in chunks in a separate table. I tested it with files of several GB of size (on a different app 😉) Obviously, the filesystem is also fine, as long as you don't want to scale out - then it gets a bit annoying imho.

**Comment by madduck** [comment]
Well, soon I'll be dealing with multi-gigabyte documents, so maybe I'll need to look into "proper streaming".

And yes, sure, chunks in PostgreSQL works, but it doesn't de-duplicate blocks.

**Comment by eikek** [comment]
Yes, there is no golden bullet. I  know people have used zfs and postgres, but I'm not sure about the implications with deduplication on.

**Comment by ahanselka** [comment]
I use Postgres with ZFS compression, which works amazingly well with Postgres for me. I haven't tried de-dupe yet though since unfortunately the prevailing opinion on the internet seems to be generally "don't".

**Comment by eikek** [comment]
Thanks for sharing @ahanselka - interesting to hear!
